File: src/main/java/util/state_machine/StateMachine.java
package util.state_machine;

import rlbotexample.dynamic_objects.DataPacket;

public class StateMachine implements Behaviour, Debuggable {
    private State state;
    private State nextState;

    public StateMachine(State initState) {
        state = null;
        nextState = initState;
    }

    public void exec(DataPacket input) {
        if(nextState != state) {
            nextState.start(input);
        }

        state = nextState;
        state.exec(input);
        nextState = state.next(input);

        if(nextState != state) {
            state.stop(input);
        }
    }

    public void debug(DataPacket input) {
        state.debug(input);
    }
}

File: src/main/python/AppMonitor/util/state_machine/state_machine.py
from typing import TypeVar, Generic

from overrides import overrides

T = TypeVar('T')


class State(Generic[T]):
    def start(self, param: T):
        pass

    def stop(self, param: T):
        pass

    def exec(self, param: T):
        pass

    def next(self, param: T):
        return self


class StateMachine(State[T]):
    def __init__(self, init_state):
        self.state = None
        self.nextState = init_state

    @overrides
    def exec(self, param: T):
        if self.nextState is not self.state:
            self.nextState.start(param)
        self.state = self.nextState
        output = self.state.exec(param)
        self.nextState = self.state.next(param)
        if self.nextState is not self.state:
            self.state.stop(param)

        return output
